README.TXT
README.TXT

E M S C A T T E R  v0.3 -- Initial Release

This is the initial release of this code, "emscatter" 

This program is provided in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. This program 
makes use of unmodified GPL and (more weakly) publicly licensed software, which
is not included here. 


0) PREAMBLE     
-----------

The EMSCATTER code calculates the T-matrices and scattering profiles of a dielectric scatterer 
encountering a monochromatic plane wave. It is a direct implementation of the method detailed in 
the paper 

B.R. Johnson,  "Invariant imbedding T matrix approach to electromagnetic scattering", 
Applied Optics, vol. 27 (1988) p. 4861


The code is written in C++. It has two external dependencies, the Eigen Matrix/ Linear Algebra
package and the popular Boost routines.

http://eigen.tuxfamily.org
http://www.boost.org


The matrix classes/representations and linear solve routines provided by Eigen are used throughout 
emscatter. One nice feature of both the Boost and Eigen  packages (at least, the parts we are using)  
is that they are provided entirely in header (*.h,*.hpp) files. They are  therefore not compiled into 
separate libraries and then linked.  

  
The EMSCATTER tarball directory structure is as follows.
    	

	0.3/
	    README.TXT -- this document
	    .../bin/
	    .../src/
	    .../include/
            
               
	    

1) INSTALLATION
---------------

There is very little configuration for this initial release. The only adjustments that needed are in the
 simple makefile,  

 	./src/Makefile


basically setting the compiler and related flags

	a) C++ compiler :              CPP = /usr/bin/g++ 
	b) C++ compile flags :         CPPCMPFLAGS = -c -O3

and the location of boost and Eigen headers, e.g.

	c) BOOST_INCDIR_        =   -I../boost/
	d) EIGEN_INCDIR_        =   -I../Eigen/
	
	

Run "make" at the prompt

     $> make   


The code should compile without issue and copy the executable "emscatter" to the directory 
EMSCATTER/bin/. 

The default dielectric structure compiled into the code is the luneberg lens. 

The code uses "long double" (64 bit) for the real floating point type (as default). This is defined in the "basic_types.hpp" header. 



2) RUNNING THE CODE
-------------------

Before the running the code, one needs to define the dielectric structure and to setup an input file.

       a) Dielectric structure: For this release, the dielectric structure needs to be defined at compile time
          and compiled into the code. The routine defining the dielectric structure is "dielectric.cpp". The prototype is
	  
	  std::complex<REAL_TYPE> dielectric(const REAL_TYPE &zc, const REAL_TYPE &phi,const REAL_TYPE &theta,const REAL_TYPE &r)
	 
	  in plain speak
    
          epsilon = dielectric(zshift,phi,theta,r)

          All arguments to the dielectric function are inputs. zshift allows one to shift the origin of the scatterer
	  along the z-axis. For this release, please only use 0.0. The shift has been tested, but special instructions are needed for 
          the input file, and those will be provided in the next release. phi, theta, and r are just the spherical coordinates 
	  of the point in the dielectric to be evaluated. The function returns a complex dielectric value at this point. 
	  The function should also return ONE (vacuum dielectric) outside the dielectric structure.

          A number of examples for dielectric functions are provided in the ./Examples/ directory. Any of these examples
	  can be copied to the src directory as dielectric.cpp. We then need to recompile. :(

	
       b) Input file: The name of the input file for the code is "emsc_input.in". Lines beginning with a '#' (comment line)
          or blank lines are ignored. Each line with valid input should form a (key,value) pair. Eack keyword should be given 
	  a single value. At the moment, the RunTitle  cannot contain spaces. 
	  
	  
	  i) Valid keywords with example value. (Keywords are case sensitive, the ':' is considered part of the keyword).
	  

	 	# INPUT FILE : emsc_input.in
		# The RunTitle is just a string that describes the run.
		RunTitle: Prolate_spheroid__c=3__b=0.2756644477109__a=2*b

                # The shift of the scatterer away from the origin along the z-axis 
		Zshift:   0.0           
		
		# Wavelength of the incoming plane wave 
		Wavelength:   1.0

                # Definition of integration region
		# Radial integration
		# Number of points for integration along radial direction
                Radius_NumPoints:      160
                # Integrate from Radius Start to Radius_End
		# Radius_Start should not be zero!!! (see below)
		# Radius_End should be at or just beyond edge or greatest extent of scatterer
 		Radius_Start:          0.0004
                Radius_End:            0.5513289
		#
		#		
		# Theta  integration, for zshift=0 this would normally be 0 to PI
		Theta_NumPoints:       240
		Theta_Start:           0.000000000000000
		Theta_End:             3.141592653589793
		#
		#
		# Phi  integration,   for zshift=0 this would normally be 0 to 2*PI
		Phi_NumPoints:         38
		Phi_Start:             0.000000000000000
		Phi_End:               6.283185307179586
		#
		# Max_L, the number of spherical harmonics used for the calculation
		Max_L:                 14
		#
		# Symmetry, one of two valid values: 'axial' or 'none' (see below)
		Symmetry:              axial


	  ii) The Radius_Start value cannot be set exactly to zero, since the Hankel functions used are
	      singular at r=0. We start with a value just off of zero, and this works surprisingly well. It can
              be shown that the T matrices --> 0 as r->0 for well behaved dielectrics. It is surprising however that 
              asymptotic expressions do not seem to be needed.

	  iii) The angular momentum components included in a given calculation are given by
	  
	  l: 1<= l <= Max_L
	  m: 0<= m <= l
	  p: 0<= p <= 1, for m => 1 ; p =0 for m=0
	  
	  If Symmetry: is set to 'none', all these components are included, there are 
	  Max_L *(Max_L +2) components. 
	      
	  If Symmetry: is set to 'axial' , then for a plane wave scatterer, the only angular momentum components
 	  that need to be included in the calculation are those for m=1. There are therefore only Max_L components 
	  in this case.    
	    
        


To run the code, copy the input file to the ./bin directory where the executable is located and type

   $> ./emscatter

 

3) OUTPUT
---------

For this release (0.3), when the code is run it prints the RunTitle to stdout, and then a statement like

T cycle ri= n 

where n is the radial integration step 0 <= n < Radial_NumPoints. Finally, it prints a statement at the end 
of the run to let you know it has completed.

All other output is written to the output file "emsc_output.out". The format of the file is as follows:



    emsc_output.out
    =======================================================================================================
    
    Input Parameters -- prints all of the input parameters in the "emsc_input.in" file
    
    T-matrices    ------------------- T_{n,np} , Tmatrices are 2x2 matrices subscripted by 2 
                                                 angular multi-indices
    n     =  l   m   p   
    np    =  lp  mp  pp
    
   ( T11     T12 )
   ( T21     T22 )_{n,np}    
 
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
	
   a & b coefficients --------------------------------
   l,m,p           ax             ay
   
   l,m,p           bx             by

   ---------------------------------------------------  a & coefficient definitions are given in
   							Johnson, Eqs. 108-112

   Scattering Intensity Functions -----------------------
   Theta            I1             I2

   ----------------------------------------------------- These intensity functions are only well defined for and only computed
                                                         for axially symmetric problems.
   
   Completed Run:Normal Exit...
   =======================================================================================================


4) EXAMPLES
-----------

	Currently, only axially symmetric examples are included, although non-symmetric runs have been tested successfully
        against another code that implements the calculation as one large matrix solve. In addition, it was verified 
	that the T matrices possess the proper symmetry for dielectrics with symmetry and  when choosing Symmetry: 'none'.
	Along with sample output, plots of the scattering intensity functions are included and may be compared to those of 
	Johnson. (The agreement with his results is excellent.)	Non-axially symmetric problems take much longer to run, obviously.

	
	./Examples/Johnson/Sphere    -- Sphere of constant dielectric
	./Examples/Johnson/Luneburg  -- Sphere with quadratic luneburg dielectric, eps = 2 - (r/r_{0})^{2}
	./Examples/Johnson/ProlateSpheroid  -- Results for 7 different prolate spheroids 
	./Examples/Cylinder/         -- Results for 2 different cylinders, radius=0.5, semi-height = 0.5,1.5 


	
	


 

 

  


CHANGE_LOG.TXT
2014-03-03;  Jfude
	     
	     *emscatter:
	     Initial Release


src/scatter.cpp
/**
 *  scatter.cpp -- EMScatter Eigen Full -- 
 *              -- Scattering code -  uses Eigen matrix package and solves the scattering problem via
 *                 interating the Tmatrix recursion relations. 
 *                 See B.R. Johnson,  "Invariant imbedding T matrix approach to
 *                 electromagnetic scattering", Applied Optics, vol. 27 (1988) p. 4861
 **/
#include<iostream>
#include<stdio.h>
#include<vector>
#include<fstream>
#include<complex>
#include<math.h>
#include<basic_types.hpp>
#include<coordParam.hpp>
#include<readInputParams.hpp>
#include<AngularMultIndex.hpp>
#include<createAngIndices.hpp>
#include<getTrapGrid.hpp>
#include<allocate.hpp>
#include<initializeT.hpp>
#include<computeBessel.hpp>
#include<computeHankel.hpp>
#include<computeGreen.hpp>
#include<computeY.hpp>
#include<computeC0.hpp>
#include<computeC1.hpp>
#include<computeD0.hpp>
#include<computeD1.hpp>
#include<computeT.hpp>
#include<computeQdd.hpp>
#include<computeQ12.hpp>
#include<computeMatrixMax.hpp>
#include<restartTfromFile.hpp>
#include<saveTtoFile.hpp>
#include<getGaussGrid.hpp>
#include<compute_ab.hpp>
#include<computeScatIntFunctions.hpp>
#include<Eigen/Dense>

int main() 
{

  // Define Numerical Constants
  const REAL_TYPE ZERO    = (REAL_TYPE)0.0000000000000000000000000000000L;
  const REAL_TYPE TWO     = (REAL_TYPE)2.0000000000000000000000000000000L;
  const REAL_TYPE PI      = (REAL_TYPE)3.1415926535897932384626433832795028842L;
  const REAL_TYPE TWOPI   = TWO*PI;
  const REAL_TYPE FOURPI  = TWO*TWOPI;
  
  // Declare the output filename and open
  std::string outFileName="emsc_output.out";
  std::string inFileName ="emsc_input.in";


  // Define most commonly used Matrix type
  typedef Eigen::Matrix<REAL_TYPE,Eigen::Dynamic,Eigen::Dynamic>                EMatrix;
  typedef Eigen::Matrix<std::complex<REAL_TYPE>,Eigen::Dynamic,Eigen::Dynamic> ECMatrix;
  
  /*~~~~~~~~~~  Declare Input Parameters ~~~~~~~~~~~~~~~~~~~~~~*/
  std::string runTitle;       // string descriptor for a specific run 
  REAL_TYPE zshift;           // shift of origin of scattering region along z-axis
  REAL_TYPE wavelength;       // wavelength in incoming plane wave
  coordParam radiusParam;     // info for integration along radial direction
  coordParam thetaParam;      // info for integration along theta direction
  coordParam phiParam;        // info for integration along phi direction
  size_t maxL;                // maximum angular momentum component (L value of spherical harmonic)
  std::string symmetry;       // symmetry of calculation: axial or none
  /*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

  
  // Define other
  bool ret;          // return value
  std::ofstream out; // output stream
  
  
  // Open output file
  out.open(outFileName.c_str());
  if(!out) {
    std::cout << "Error:: Unable to open output file... "<< outFileName << std::endl;
    return 0;
  }
  
  
  //Read input file
  ret=readInputParams(runTitle,zshift,wavelength,radiusParam,thetaParam,phiParam,maxL,symmetry, inFileName, out);

  if(!ret) {
    std::cout << "Read of input file :" << inFileName << " failed." << std::endl;
    out       << "Read of input file :" << inFileName << " failed." << std::endl;
    // close output file
    out.close(); // need check?
    return 0;
  }
    

  // Write the input to the output file
  std::cout << runTitle << std::endl;  
  out << runTitle << std::endl << std::endl << std::endl;
  out << "zshift     = " << zshift     << std::endl;
  out << "wavelength = " << wavelength << std::endl;
  out << "radiusParam.nPoints  =  "  << radiusParam.nPoints << std::endl;
  out << "radiusParam.start    =  "  << radiusParam.start   << std::endl;
  out << "radiusParam.end      =  "  << radiusParam.end     << std::endl;
  out << "thetaParam.nPoints   =  "  << thetaParam.nPoints  << std::endl;
  out << "thetaParam.start     =  "  << thetaParam.start    << std::endl;
  out << "thetaParam.end       =  "  << thetaParam.end      << std::endl;
  out << "phiParam.nPoints     =  "  << phiParam.nPoints    << std::endl;
  out << "phiParam.start       =  "  << phiParam.start      << std::endl;
  out << "phiParam.end         =  "  << phiParam.end        << std::endl;
  out << "maxL                 =  "  << maxL                << std::endl;
  out << "symmetry             =  "  << symmetry            << std::endl;
  out << std::endl << std::endl;
  //


  //
  // const bool saveT = false;
  // const std::string saveTFile    = "restart.dat";
  // const bool restartT = false;
  // const std::string restartTFile = "restart.dat"; 
  //
  
  
  /* Define the ang momentum index */
  std::vector<AngularMultIndex> angInd;
  if(!createAngIndices(maxL,angInd,symmetry,out)) return 0;
  
  const size_t nsize  = angInd.size();
  const REAL_TYPE k = TWOPI/wavelength;
  

  /* ~~~~~~ Define and allocate matrix container objects upfront ~~~~~~~~~~~~~~~~~~~~ */

  // Bessel Matrix, diagonal in (l,l') indices for a given r. 
  // Fundamental matrix is 3x2. Index layout: J[l](,)
  std::vector<EMatrix>   J;
  
  // Hankel Matrix, diagonal in (l,l') indices for a given r. 
  // Fundamental matrix is 3x2. Index layout: H[l](,)
  std::vector<ECMatrix>  H;

  // Radial greens function matrix composed from products of
  // J and H. G is diagonal in (l,l'), this is diagonal component of g(r,r')
  // for a given r
  // Fundamental matrix is 3x3, Index layout: G[l](,) 
  std::vector<ECMatrix>  G;
  
  // U matrix: Fundamental matrix is 3x3, Index layout 
  // U[n][n'](,)
  //std::vector<std::vector<EMatrix> > U;
  
  // C0 = w*U at a given r, raw matrix (3*nsize,3*nsize)
  //ECMatrix C0;
  // can maybe get away with EMatrix
  ECMatrix C0;
  
  // C1 = 1 - C0*g, raw matrix (3*nsize,3*nsize)
  ECMatrix C1;

  // Q matrix, raw matrix, (2*nsize,2*nsize) solution of C1*Q=C0;
  ECMatrix Q;
  
  // T matrix, raw matrix (2*nsize,2*nsize)
  ECMatrix T;

  // Q11 = Jt * Q * J  matrix nsize2 x nsize2 
  // Q12 = (Jt * Q * H) + I  matrix nsize2 x nsize2 
  // Q22 = Ht * Q * H  matrix nsize2 x nsize2 
  ECMatrix Q11,Q12,Q22; 

  // D0 = T(r_{i-1}) *[I + Q12.transpose())], nsize2 x nsize2
  ECMatrix D0;

  // D1 = I - T(r_{i-1})*Q22
  ECMatrix D1;

  // q matrix as in D1*q=D0;, (2*nsize, 2*nsize)
  ECMatrix q;

  // check matrices
  ECMatrix QCcheck,QDcheck;
  
  // Y matrices
  std::vector<std::vector<std::vector<EMatrix> > > Y;

  // a and b coeffs
  ECMatrix a_coeff,b_coeff;
  
  size_t i,j,pp;
  
  
  
  /* Compute radial angular grids */
  //should wrap allocations in try
  
  std::vector<REAL_TYPE>   r(radiusParam.nPoints);
  std::vector<REAL_TYPE> w_r(radiusParam.nPoints);
  std::vector<REAL_TYPE>   theta(thetaParam.nPoints);
  std::vector<REAL_TYPE> w_theta(thetaParam.nPoints);
  std::vector<REAL_TYPE>   phi(phiParam.nPoints);
  std::vector<REAL_TYPE> w_phi(phiParam.nPoints);


  if(!getTrapGrid( r    ,w_r    ,radiusParam.nPoints,radiusParam.start  ,radiusParam.end))           return 0;
  if(!getGaussGrid(theta,w_theta,thetaParam.nPoints ,thetaParam.start,thetaParam.end,out))          return 0;
  if(!getGaussGrid(phi  ,w_phi  ,phiParam.nPoints   ,phiParam.start  ,phiParam.end,out))            return 0;
    
  
  // Write grids to output file
  /*
  std::cout << "rsize = " << r.size() << std::endl;
  for(i=0;i<nTheta;i++) {
    std::cout << theta[i] << "   "<< w_theta[i] << std::endl;
  }
  for(i=0;i<nPhi;i++) {
    std::cout << phi[i] << "   "<< w_phi[i] << std::endl;
  }
  */
  
  
  
  // Allocate other matrices used in the T cycle loop
  if(!allocate("J Bessel Matrix",         J,maxL,3,2,out)) return 0;
  if(!allocate("H Hankel Matrix",         H,maxL,3,2,out)) return 0;
  if(!allocate("G Greens Function Matrix",G,maxL,3,3,out)) return 0;
  if(!allocate("C0  Matrix",      C0,3*nsize,3*nsize,out)) return 0;
  if(!allocate("C1  Matrix",C1,3*nsize,3*nsize,out))   return 0;  
  if(!allocate("Q  Matrix  ",Q,  2*nsize,2*nsize,out)) return 0;
  if(!allocate("Q11  Matrix",Q11,2*nsize,2*nsize,out)) return 0;
  if(!allocate("Q12  Matrix",Q12,2*nsize,2*nsize,out)) return 0;
  if(!allocate("Q22  Matrix",Q22,2*nsize,2*nsize,out)) return 0;
  if(!allocate("D0   Matrix",D0 ,2*nsize,2*nsize,out)) return 0;
  if(!allocate("D1   Matrix",D1 ,2*nsize,2*nsize,out)) return 0;  
  if(!allocate("q    Matrix",q,2*nsize,2*nsize,out))   return 0;
  if(!allocate("T    Matrix",T,2*nsize,2*nsize,out))   return 0;
  if(!allocate("a_coeff Matrix",a_coeff,2,nsize,out))  return 0;
  if(!allocate("b_coeff Matrix",b_coeff,2,nsize,out))  return 0;
  // Qcheck matrices
  if(!allocate("QC check Matrix",QCcheck,3*nsize,3*nsize,out))   return 0;
  if(!allocate("QD check Matrix",QDcheck,2*nsize,2*nsize,out))   return 0;
  
  if(!allocate("Y matrix",Y,phiParam.nPoints,thetaParam.nPoints,nsize,3,3,out))  return 0;
 

  /* ~~~~~~~~~~~~~~~~~~~  Loop over radial increments  ~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  // Might be able to combine J and H compute, like compute<sph_bessel>                  

  // Iniitialize T matrix here (for restarting)
  /******************************************************************************
  if(restartT) {
    std::cout << "Restarting T matrix from File"<<std::endl;
    if(!restartTfromFile(r_begin,T,restartTFile)) return 0;
    r_begin += r_begin_shift;
  }
  else {
    // initialize at zero
    std::cout << "Initializing Tmatrix at zero.."<< std::endl;
    
    // reset rbegin here
  }
  *******************************************************************************/
  
  //Initialize T
  if(!initializeT(T)) {
    // check print failure
    out.close();
    return 0;
  }
  
  // Compute Y matrices
  if(!computeY(Y,angInd,maxL,theta,w_theta,phi,w_phi,out)) {
    // check print failure
    out.close();
    return 0;
  }

 
  bool computeError=false;
  REAL_TYPE qMax;
  int jj;
  
  
  /*~~~~~~~~~~~~~~~~~~~~~~~~~~~ T  C y c l e   L o o p ~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  
  for(i=0;(!computeError) && i<radiusParam.nPoints;i++) {
    
    // COMPUTE BESSEL MATRIX
    if(!computeBessel(J,k,r[i],maxL,out))    {
      computeError=true;continue;
    }
    
    // COMPUTE HANKEL MATRIX
    if(!computeHankel(H,k,r[i],maxL,out))    {
      computeError=true;continue;
    }
    
    // COMPUTE GREENS FUNCTION MATRIX
    if(!computeGreen(G,J,H,k,out)) {
      computeError=true;continue;
    }

    // Compute C0 Matrix , C0 = w_{n}*U(r_{n}), C0 is my notation 
    // w*U is from Eq. (84) in Johnson
    if(!computeC0(C0,Y,angInd, zshift, phi,w_phi,theta,w_theta,r[i],w_r[i],k,out)) {
      computeError=true;continue;
    }    
    
    
    // Compute C1 matrix C1 = 1 - C0*g = 1 - w*U*g from Eq. (84)
    if(!computeC1(C1,C0,G,angInd,out)) {
      computeError=true;continue;
    }
    
    
    // Solve Eq. (84)
    // need to catch specific exceptions
    try {
      Q  = C1.partialPivLu().solve(C0);
    }
    catch(...) {
      std::cout << "Problem solving for Q" << std::endl;
      out << "Problem solving for Q" << std::endl;
      computeError=true;continue;
    }
    
    
    // Compute solution difference
    //QCcheck.noalias() = C1*Q - C0;
    //qMax = computeMatrixMax(QCcheck);
    //std::cout << "max_norm(C1*Q - C0)  = " << qMax << std::endl;
    
    
    // Compute Q11, Eq. 85
    if(!computeQdd<EMatrix> (Q11,Q,J,  angInd,k,out)) {
      computeError=true;continue;
    }
    // Compute Q12, Eq. 86
    if(!computeQdd<ECMatrix>(Q22,Q,H,  angInd,k,out)) {
      computeError=true;continue;
    }
    // Compute Q22, Eq. 87
    if(!computeQ12(Q12,Q,J,H,angInd,k,out)) {
      computeError=true;continue;
    }
   
    
    // Compute D0 = T * (1 + Q12)
    if(!computeD0(D0,T,Q12,out)) {computeError=true;continue;} 
    // Compute D1 = 1 - T*Q22
    if(!computeD1(D1,T,Q22,out)) {computeError=true;continue;}
    
    
    // By combining Eqs. 94 & 95, one gets D1*q = D0
    // 
    // Here we solve this equation for q
    // need to catch all the exceptions
    try {
      q  = D1.partialPivLu().solve(D0);
    }
    catch(...) {
	std::cout << "Problem solving for Q..." << std::endl;
	return 0;
    }   
  
    
    //std::cout << "Computed q  Matrix"<< std::endl;
    //QDcheck.noalias() = D1*q - D0;
    //qMax = computeMatrixMax(QDcheck);
    //std::cout << "max_norm(D1*Q - D0)  = " << qMax << std::endl;

    
    
    // Compute next value of T, this is Eq. 97
    T.noalias() = Q11 + q + Q12*q;
    
    std::cout << "T cycle ri = " << i << std::endl;
    
  }

  // need to take down most memory here

  
  if(computeError) {
    out.close();
    return 0;
  }

  // Save T matrix if requested
  // if(saveT) {
  //  if(!saveTtoFile(r[nRadPts-1],T,saveTFile)) return 0;
  //}
  
  
  // Print T matrices to output file

  out << "~~~~~~~~~~~~~~~~~~  T  M a t r i c e s  ~~~~~~~~~~~~~~~~~~~~~~~~~~" << std::endl;   
  for(i=0;i<nsize;i++) {
    size_t lfc = 2*i;
    for(j=0;j<nsize;j++) {
      size_t rfc = 2*j;
      
      out << "n  =   " << angInd[i].l << "   " << angInd[i].m << "   " << angInd[i].p << std::endl;
      out << "np =   " << angInd[j].l << "   " << angInd[j].m << "   " << angInd[j].p << std::endl;
      out << T.block(lfc,rfc,2,2) << std::endl;
      out << std::endl << std::endl;
    }
  }
  out << "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~" << std::endl;   
  out << std::endl << std::endl;  
  
  // need to flush the output

  
  // Compute a & b coefficients (see Johnson Eqs. 108 - 112)
  if(!compute_ab(a_coeff,b_coeff,T,angInd)) {
    out.close();
    return 0;
  }
    
  
  // Print a & b coefficients
  out << "a & b coefficients --------------------------------" << std::endl;
  out << "n           ax             ay" << std::endl;
  for(i=0;i<nsize;i++) {
    out << angInd[i].l << angInd[i].m << angInd[i].p << "   " << a_coeff(0,i)  << "    " << a_coeff(1,i) << std::endl;
  }
  
  out << "n           bx             by" << std::endl; 
  for(i=0;i<nsize;i++) {
    out << angInd[i].l << angInd[i].m << angInd[i].p << "    "<<  b_coeff(0,i)  << "    " << b_coeff(1,i) << std::endl;
  }
  out << "---------------------------------------------------" << std::endl;
  out << std::endl << std::endl;  
  


  // For axially symmetric problems  
  std::vector<REAL_TYPE> Itheta,I1,I2;
  const size_t nItheta = 1000;
  
  REAL_TYPE p11;
  if(symmetry=="axial") {
    Itheta.resize(nItheta);
    I1.resize(nItheta);I2.resize(nItheta); // should be allocated before T cycle loop, checked
    if(!computeScatIntFunctions(I1,I2,Itheta,angInd,a_coeff,b_coeff)) {
      out.close();
      return 0;
    }
    
    out << std::endl << "Scattering Intensity Functions -----------------------" << std::endl;
    out << "Theta            I1             I2              P11  " << std::endl;
    
    
    for(i=0;i<Itheta.size();i++) {   
      out.width(15);
      out.precision(8);
      p11 = (I1[i] + I2[i])/TWO;
      out << Itheta[i] << "         " << I1[i] << "         " << I2[i] << "        " << p11 << std::endl;
    }
    
  }

  
  out       << "Completed Run:Normal Exit..." << std::endl;
  std::cout << "Completed Run:Normal Exit..." << std::endl;
  
  out.close();
  return 0;
  
}
